mem.go

```diff
@@ -571,40 +573,51 @@ func (e *Encoder) Encode(m *Message) error {
 	if nsegs == 0 {
 		return errMessageEmpty
 	}
-	sizes, err := m.segmentSizes()
-	if err != nil {
-		return err
-	}
-	maxSeg := uint32(nsegs - 1)
-	hdrSize := streamHeaderSize(maxSeg)
-	if uint64(cap(e.hdrbuf)) < hdrSize {
-		e.hdrbuf = make([]byte, hdrSize)
+	if int64(cap(e.bufs)) < 1+nsegs {
+		e.bufs = make([][]byte, 1+nsegs)
 	} else {
-		e.hdrbuf = e.hdrbuf[:hdrSize]
+		e.bufs = e.bufs[:1+nsegs]
 	}
-	marshalStreamHeader(e.hdrbuf, sizes)
-	if err := e.write(e.hdrbuf); err != nil {
-		return err
+	if int64(cap(e.sizes)) < nsegs {
+		e.sizes = make([]Size, nsegs)
+	} else {
+		e.sizes = e.sizes[:nsegs]
 	}
 	for i := int64(0); i < nsegs; i++ {
 		s, err := m.Segment(SegmentID(i))
 		if err != nil {
 			return err
 		}
-		if err := e.write(s.data); err != nil {
-			return err
+		n := len(s.data)
+		if int64(n) > int64(maxSize) {
+			return errSegmentTooLarge
 		}
+		e.sizes[i] = Size(n)
+		e.bufs[1+i] = s.data
 	}
-	return nil
+	maxSeg := uint32(nsegs - 1)
+	hdrSize := streamHeaderSize(maxSeg)
+	if uint64(cap(e.hdrbuf)) < hdrSize {
+		e.hdrbuf = make([]byte, hdrSize)
+	} else {
+		e.hdrbuf = e.hdrbuf[:hdrSize]
+	}
+	marshalStreamHeader(e.hdrbuf, e.sizes)
```

Comment:
TBH, I think I want to nuke `marshalStreamHeader`.  It's adding an allocation for not much clarity.  Let's inline it here (and define a new helper, `appendUint32`):

```go
if uint64(cap(e.hdrbuf)) < hdrSize {
  e.hdrbuf = make([]byte, 0, hdrSize)
}
e.hdrbuf = appendUint32(e.hdrbuf[:0], maxSeg)
for i, buf := range e.bufs {
 e.hdrbuf = appendUint32(e.hdrbuf, uint32(Size(len(buf))/wordSize))
}
if len(e.hdrbuf) % int(wordSize) != 0 {
  e.hdrbuf = appendUint32(e.hdrbuf, 0)
}
```

Interestingly, I think that the current `Encoder` has a padding bug in this regard.  Can you add a test where a 3-segment message is encoded then a 2-segment message is encoded, and then assert that the 2-segment message's header properly contains zero-padding (instead of garbage)?